design-documents/graph-ql/framework/async-mutations.md

```diff
@@ -0,0 +1,281 @@
+}
+```
+
+#### Request
+```graphql 
+query {
+    queue($operataionIdentifier: [String!]) {
+        identifier
+        __typename__
+        ... on PersistOperationCompleted {
+            changedEntities {
+                type
+                identifier
+            }
+        }
+    }
+}
+```
+#### Response
+```json
+{
+   "queue":[
+      {
+         "identifier":"3d8d487165a5653733e3a3e8bd46783e46b9a58d",
+         "__typename__":"PersistOperationCompleted",
+         "changedEntities":[
+            {
+               "Cart":[
+                  "29f77a662a7f4d019d5160a9dbe4b83067a75f68"
+               ]
+            }
+         ]
+      }
+   ]
+}
+```
+#### Custom
+
+A developer can specify custom operation status, which may aggregate success result with output information.
+For instance, `CartSaveOperationCompleted` state may, return a corresponding cart as with query.
+```graphql schema
+type CartSaveOperationCompleted implements OperationStateInterface {
+     identifier: String!
+     cart: Cart!
+}
+```
+For such cases, a client should be aware of the operation and result that it expects to receive.
+This is kind of trade-off between generalized approach and numbers of request to the server.
+Such an approach may help to achieve better performance but definitely will cost more in development and testing.
+#### Request
+```graphql 
+query {
+    queue($operataionIdentifier: [String!]) {
+        identifier
+        __typename__
+        ... on CartSaveOperationCompleted {
+            cart {
+                increment_id
+                items {
+                    sku
+                    qty
+                }
+            }
+        }
+    }
+}
+```
+#### Response
+```json
+{
+   "queue":[
+      {
+         "identifier":"3d8d487165a5653733e3a3e8bd46783e46b9a58d",
+         "__typename__":"PersistOperationCompleted",
+         "cart":{
+            "increment_id":"AA000000001",
+            "items":[
+               {
+                  "sku":"simple-product-1",
+                  "qty":1
+               },
+               {
+                  "sku":"simple-product-2",
+                  "qty":2
+               }
+            ]
+         }
+      }
+   ]
+}
+```
+
+### Mimicking async mutation
+
+Actually, with this proposal, we will have to change application workflows to support asynchronous communication. With the great benefit for the future system will face some challenges during the transition period.
+Async operations cause the domino effect. There are not enough to change the single mutation, all operation should be rewritten at some moment.
+Taking to account backward compatibility policies the migration logic should be defined in advance.
+So, the system should be proactive and provides a possibility to write a client application which will be resilient to mutations mode switch.
+
+To achieve this goal we can wrap the existing synchronous operations with `OperationCompleted` state.
```

Comment:
This is non-standard approach, which main benefit is ability to simplify client refactoring when they need async GraphQL for performance improvement. The main drawback is that the interface becomes complicated for everyone now, even for those who do not need async GraphQL.